In [444]:
# import all libararies that we will need in this project 
import pandas as pd
import datetime as dt
import numpy as np
import matplotlib.pyplot as plt 
import  seaborn as sns
%matplotlib inline
import json
import os
import requests
import tweepy
from IPython.display import HTML

<h3> 1- Gathring data :

In [445]:
#to read the data which we will use it
twitter_arch = pd.read_csv('twitter-archive-enhanced.csv')

In [446]:
twitter_arch.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193 NaN                    NaN                    
1  892177421306343426 NaN                    NaN                    
2  891815181378084864 NaN                    NaN                    
3  891689557279858688 NaN                    NaN                    
4  891327558926688256 NaN                    NaN                    

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                                                               source  \
0  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
3  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
4  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                                 text  \
0  This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU                                                        
1  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   
2  This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB                    
3  This is Darla. She commenced a snooze mid meal. 13/10 happens to the best of us https://t.co/tD36da7qLQ                                                              
4  This is Franklin. He would like you to stop calling him "cute." He is a very fierce shark and should be respected as such. 12/10 #BarkWeek https://t.co/AtUZn91f7f   

   retweeted_status_id  retweeted_status_user_id retweeted_status_timestamp  \
0 NaN                  NaN                        NaN                         
1 NaN                  NaN                        NaN                         
2 NaN                  NaN                        NaN                         
3 NaN                  NaN                        NaN                         
4 NaN                  NaN                        NaN                         

                                                                                                                     expanded_urls  \
0  https://twitter.com/dog_rates/status/892420643555336193/photo/1                                                                   
1  https://twitter.com/dog_rates/status/892177421306343426/photo/1                                                                   
2  https://twitter.com/dog_rates/status/891815181378084864/photo/1                                                                   
3  https://twitter.com/dog_rates/status/891689557279858688/photo/1                                                                   
4  https://twitter.com/dog_rates/status/891327558926688256/photo/1,https://twitter.com/dog_rates/status/891327558926688256/photo/1   

   rating_numerator  rating_denominator      name doggo floofer pupper puppo  
0  13                10                  Phineas   None  None    None   None  
1  13                10                  Tilly     None  None    None   None  
2  12                10                  Archie    None  None    None   None  
3  13                10                  Darla     None  None    None   None  
4  12                10                  Franklin  None  None    None   None

In [447]:
twitter_arch.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

<h3>2- Tweet images prediction :
   

In [448]:
# download dataset programatically:
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response =requests.get(url)
response


ConnectionError: HTTPSConnectionPool(host='d17h27t6h515a5.cloudfront.net', port=443): Max retries exceeded with url: /topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x00000234168AD6A0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))

In [ ]:
#create a folder for image predictions data
folder_name = 'image_predictions'
if not os.path.exists(folder_name):
    os.makedirs(folder_name)

In [ ]:
#Write image predictions data to a file in our folder
with open (url.split('/')[-1],mode = 'wb') as file:
      file.write(response.content)
#reading the TSV file :
image_prediction = pd.read_csv('image-predictions.tsv', sep='\t' )

In [ ]:
image_prediction.head()

In [ ]:
#Exploration of data types and missing fields.
image_prediction.info()

<h3> 3- Twitter API


In [ ]:
from tweepy import OAuthHandler
from timeit import default_timer as timer
from credentials import *
from time import sleep
# Query Twitter API for each tweet in the Twitter archive and save JSON in a text file
# These are hidden to comply with Twitter's API terms and conditions

consumer_key = 'YOUR CONSUMER KEY'
consumer_secret = 'YOUR CONSUMER SECRET'
access_token = 'YOUR ACCESS TOKEN'
access_secret = 'YOUR ACCESS SECRET'

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth)



<h3> #read in the JSON data from the text file, and save to a DataFrame


In [ ]:
#https://towardsdatascience.com/flattening-json-objects-in-python-f5343c794b10
api_data = []

#read the created file
with open("tweet_json", "r") as f:
    for line in f:
        try: 
            tweet = json.loads(line)
            #append a dictionary to the created list            
            api_data.append({
                "tweet_id": tweet["id"],
                "retweet_count": tweet["retweet_count"],
                "favorite_count": tweet["favorite_count"],
                "retweeted": tweet["retweeted"],
                "display_text_range": tweet["display_text_range"]                
            })               
                
            #tweet["entities"]["media"][0]["media_url"]
        except:
            print("Error.") 
            
df_api = pd.DataFrame(api_data, columns = ["tweet_id", "retweet_count", "favorite_count", "retweeted", "display_text_range"])
df_api.head()

In [ ]:
twitter_arch.head(1)


In [ ]:
df_api.head(1)

In [ ]:
image_prediction.head(1)

<h3>Data Assessing :</h3>
Now we are going to look for issues with the quality or the tidiness of the data and its dimensions.


1)Let's first look for missing data.



In [ ]:
fig, ax = plt.subplots(figsize = (20,7))
ax = sns.heatmap(twitter_arch.isnull(), vmin=0, vmax = 1)



As we can see, there is a lot of missing data in the columns about the reply and the retweeted status. Since we only want original posts with images, we have to drop them later - the missing data in the "expanded_urls" column will also disappear with that cleaning operation.

In [ ]:
twitter_arch.sample(3)


During the visual assessment we can observe that:

<ol><li> not all tweets could be classified as doggo, floofer, pupper or puppo and all columns contain "None"
<li>the source contains unnecessary HTML code
<li>there is the name "None" in the name column</ol>

In [ ]:
twitter_arch.info()


Also we found the datatypes are incorrect:

<ul><li>tweet_id should be a str
<li>timestamp - columns should be datetime objects
<li>Now let's see how many wrong names we can find.</ul>

In [ ]:
twitter_arch.name.value_counts()


also, name column contains wrong names like "None", "a", "the", "an".

In [ ]:
twitter_arch[twitter_arch.duplicated()]


good, we don found any duplicates in this data, so the number of unique tweet_ids should be the length of the data frame (twitter_arch)



In [ ]:
twitter_arch.tweet_id.nunique()


good .  we will create a copy of the "twitter_arch" dataframe for further assessing. I want to see in how many cases there were no classification of dog_size possible via text processing.

In [ ]:
tw_arch_copy = twitter_arch.copy()


In [ ]:
#returns true if there is no dog classification in any of the columns
tw_arch_copy[["doggo","floofer","pupper","puppo"]].apply(lambda x: True if
    (x[0] == "None" and x[1] == "None" and x[2] == "None" and x[3] =="None") 
    else False, axis = 1).value_counts()

In [ ]:
tw_arch_copy["doggo"].value_counts()


In [ ]:
tw_arch_copy["floofer"].value_counts()

In [ ]:
print(tw_arch_copy["pupper"].value_counts())
print(tw_arch_copy["puppo"].value_counts())



Only for 16% of the rows the data is not missing. Now let's take a look at the ratings. 
By what we have seen so far, it looks like the ratings have always a format of 13/10 or 12/10 and so on. 
So we would expect a numerator > 10 and denominator = 10.



In [ ]:
tw_arch_copy.rating_numerator.value_counts()


 there is a wide range of numbers as rating_numerator, with a maximum of 1776.



In [ ]:
print(tw_arch_copy.query("rating_numerator == '75'").text)

This rating is correct - so a lot higher values as 12 or 13 would be valid to this rating system. But what is with the very small ones?



In [ ]:
print(tw_arch_copy.query("rating_numerator == '1'").text)


The entry 605 shows, that these tweets contain pictures that don't contain any dogs. 
Also in the entry 2335 the rating got extracted wrongly (misinterpreted the 1/2 of 3 1/2 as the rating)

In [ ]:
print(tw_arch_copy.query("rating_numerator == '0'").text)


In [ ]:
tw_arch_copy.rating_denominator.value_counts()


In [ ]:
print(tw_arch_copy.query("rating_denominator == '120'").text)


In [ ]:
print(tw_arch_copy.query("rating_denominator == '0'").text)


In [ ]:
print(tw_arch_copy.query("rating_denominator == '170'").text)


same problems as for the numerator. Multiple Dogs or multiple occurences of the pattern \d+\/\d+.
Lets extract this by our self and see what we get.



In [ ]:
pattern = "(\d+(\.\d+)?\/\d+(\.\d+)?)" #we could expect an integer rating on what we saw, but maybe some floats are the case

#https://stackoverflow.com/questions/36028932/how-to-extract-specific-content-in-a-pandas-dataframe-with-a-regex
tw_arch_copy["rating"] = tw_arch_copy.text.str.extract(pattern, expand = True)[0]

#https://stackoverflow.com/questions/14745022/how-to-split-a-column-into-two-columns
tw_arch_copy[['num', 'denom']] = tw_arch_copy['rating'].str.split('/', n=1, expand=True)

In [ ]:
tw_arch_copy.rating_numerator = tw_arch_copy.rating_numerator.astype("str")
tw_arch_copy.rating_denominator = tw_arch_copy.rating_denominator.astype("str")


In [ ]:
#look for differences in the original numerator and the newe extract
tw_arch_copy["check_num"] = tw_arch_copy[["rating_numerator", "num"]].apply(lambda x: False if (x[0] != x[1]) else True, axis = 1)


In [ ]:
tw_arch_copy.check_num.value_counts()


In [ ]:
tw_arch_copy.query("check_num == False")[["rating_numerator", "num","check_num"]]


These are the differences we founded by extracting the first occurrence of the pattern , so These ratings got transformed to integers and are therefore wrong.



In [ ]:
tw_arch_copy[["rating_numerator", "num", "check_num"]].sample(15)


and with repeating it  for the denominator.

In [ ]:
tw_arch_copy["check_denom"] = tw_arch_copy[["rating_denominator", "denom"]].apply(lambda x: False if (x[0] != x[1]) else True, axis = 1)

In [ ]:
tw_arch_copy.check_denom.value_counts()


In [ ]:
tw_arch_copy.query("check_denom == False")[["rating_denominator", "denom","check_denom"]] #problem with integer, maybe also floats?


This seems like it is no problem that we have to worry about.



In [ ]:
tw_arch_copy[["rating_denominator", "denom", "check_denom"]].sample(5)

  should be to assessing  how often there are multiple occurences of the "rating pattern" in one tweet.



In [ ]:
tw_arch_copy["count"] = tw_arch_copy.text.str.count(pattern)

In [ ]:
tw_arch_copy["count"].value_counts()


In [ ]:
#show the full text
pd.set_option('display.max_columns', 500)
pd.set_option('display.max_colwidth', 4000)

tw_arch_copy[["text", "count"]].query("count != 1")

then, we can see:

this data contains retweets (as mentioned before)
sometimes there are multiple dogs/cats or else in one picture
some of these ratings are not clear

* <h3>image_prediction :<br>
Let's also begin with the missing data first.



In [ ]:
fig, ax = plt.subplots(figsize = (20,7))
ax = sns.heatmap(image_prediction.isnull(), vmin = 0, vmax = 1)


Looks fine. Now the visual assessment:



In [ ]:
image_prediction.sample(12)

We can see that:
<ul>
<li>the predicitions are sometimes lowercase, sometimes uppercase
<li>there is an underscore instead of a whitespace between the words
<li>there are rows with no prediciton of a dog (neither in 1, 2 nor 3)</ul

In [ ]:
image_prediction.info()

1- the tweet_id colum should again be string
2- The best way to find duplicates is to look at the jpg - url,If there are value counts > 1, then this data contains duplicates/retweets.

In [ ]:
image_prediction.jpg_url.value_counts().head(10)


As predicted - this data contains retweets.

In [ ]:
image_prediction[image_prediction.tweet_id.duplicated()]


In [ ]:
image_prediction[image_prediction.jpg_url == "https://pbs.twimg.com/media/CeRoBaxWEAABi0X.jpg"]


In [ ]:
twitter_arch[twitter_arch.tweet_id == 798697898615730177]


We only want Tweets with pictures which contain dogs. Let's see if there are pictures, for which the ML - Algorithm didn't predict any dogs.


In [ ]:
image_prediction.query("p1_dog == False and p2_dog == False and p3_dog == False")

In [ ]:
image_prediction.query("p1_dog == False and (p2_dog == True or p3_dog == True)")

After checking some of these pictures it gets clear, that sometimes the doggos are in the background or the pictures doesn't contain any dogs at all.



df_api

Lets repeat the same pattern for this dataset: missing data → visual assessment.



In [ ]:
fig, ax = plt.subplots(figsize = (20,7))
ax = sns.heatmap(df_api.isnull(), vmin = 0, vmax = 1)


In [ ]:
df_api.sample(5)

In [ ]:
df_api.info()


Overall it looks good for this dataset. There is no missing data - only the datatype of the tweet_id should again be a string.



Assessing Summary:

<h3>Quality</h3>
<ol>
<li>twitter_arch table:
<ul>
<li>the datatype of the id - columns is integer and should be str
<li>the datatype of the timestamp - column is object and should be datetime
<li>some of the dogs are not classified as one of "doggo", "floofer", "pupper" or "puppo" and contain all "None" instead
<li>some of the dog names are not correct (None, an, by, a, ...)contains retweets
<li>some of the ratings are not correctly extracted (mostly if there are >1 entries with the pattern "(\d+(.\d+)?\/\d+(.\d+)?)"
<li>also transforming the ratings to integer created some mistakes (there are also floats)the source column contains html code
    </ul>
2-imge_prediction table:
<ul>
<li>the datatype of the id - columns is integer and should be str
<li>contains retweets (duplicated rows in column jpg_url)
<li>there are pictures in this table that are not dogs
<li>the predictions are sometimes uppercase, sometimes lowercase
<li>also there is a "_" instead of a whitespace in the predictions
    
3-df_api table

<li>the datatype of the id - columns is integer and should be str
   </ul>
   
<h3>Tidiness</h3>
<ul>
1-df_twitter table<br>
<li>he columns doggo, floofer,pupper and puppo are not easy to analyze and should be in one column<br>
2-df_predict table<br>
<li>the prediction and confidence columns should be reduced to two columns - one for the prediction with the highest confidence (dog)<br>
3- df_api table<br>
<li>display_text_range contains 2 variables</ul>

conclusion:<br>
All three tables share the column tweet_id and should be merged together.



<h3>Data Cleaning</h3>
Cleaning steps:
<ol>
<li>Merge the tables together
<li>Drop the replies, retweets and the corresponding columns and also drop the tweets without an image or with images which don't display doggos
<li>Clean the datatypes of the columns
<li>Clean the wrong numerators - the floats on the one hand (replacement), the ones with multiple occurence of the pattern on the other (drop)
<li>Extract the source from html code
<li>Split the text range into two separate columns
<li>Remove the "None" out of the doggo, floofer, pupper and puppo column and merge them into one column
<li>Remove the wrong names of name column
<li>Reduce the prediction columns into two - breed and conf
<li>Clean the new breed column by replacing the "_" with a whitespace and make them all lowercase</ol>

<h4>1- Merge the tables together



I could clean all the tables one by one, but all of them share some cleaning needs or are dependent on each other to do so (for example removing of retweets or pictures not containig dogs). By merging them all together as first step, I can save some coding time and avoid repetition

In [ ]:
#outer join to not loose rows at first
df_master = pd.merge(twitter_arch, df_api, on = "tweet_id", how = "outer")


In [ ]:
df_master = pd.merge(df_master, image_prediction, on = "tweet_id", how = "outer")


In [ ]:
df_master_clean = df_master.copy()


In [ ]:
df_master_clean.info()

<h4> 2. Drop the replies, retweets and the corresponding columns and also drop the tweets without an image or with images which don't display doggos



Now that this is done, we have to remove the replies, retweets and the tweets withouth an image displaying a dog, because we only want original tweets with images. Let's visualize our missing data first.



In [ ]:
fig, ax = plt.subplots(figsize = (20,7))
ax = sns.heatmap(df_master_clean.isnull(), vmin = 0, vmax = 1)

In [ ]:
pd.set_option('display.max_colwidth', 50)
df_master_clean[df_master_clean["retweeted"].isnull()]


In [ ]:
#we only want the rows without an entry in "retweeted_status_id" in our master dataframe
df_master_clean = df_master_clean[df_master_clean["retweeted_status_id"].isnull()]


In [ ]:
fig, ax = plt.subplots(figsize = (20,7))
ax = sns.heatmap(df_master_clean.isnull(), vmin = 0, vmax = 1)


In [ ]:
df_master_clean.columns


In [ ]:
#check with the column from the api table, no retweets left
df_master_clean.retweeted.value_counts()

In [ ]:
#same as for the retweets, we only want the rows without an entry in "in_reply_to_status_id"
df_master_clean = df_master_clean[df_master_clean.in_reply_to_status_id.isnull()]


In [ ]:
fig, ax = plt.subplots(figsize = (20,7))
ax = sns.heatmap(df_master_clean.isnull(), vmin = 0, vmax = 1)

During the gathering process from the API there were some tweets, which got deleted by the account. We will also drop them out of our master dataframe.

In [ ]:
df_master_clean.dropna(subset = ["retweeted"], inplace = True)

In [ ]:
#drop the unneeded columns
df_master_clean.drop(["in_reply_to_status_id", "in_reply_to_user_id",
                      "retweeted_status_id", "retweeted_status_user_id", 
                      "retweeted_status_timestamp", "retweeted"], inplace=True, axis = 1)

In [ ]:
fig, ax = plt.subplots(figsize = (20,7))
ax = sns.heatmap(df_master_clean.isnull(), vmin = 0, vmax = 1)


Now we want to take a look on the "jpg_url" column and drop all the rows, which are NAN - since these are the ones without an image. To check that, we could read in the image data from the gathered API data - but I will leave this for another time.

In [ ]:
df_master_clean.dropna(subset = ["jpg_url"], inplace = True)


In [ ]:
#check if there are still duplicated images after dropping the replies and the retweets
sum(df_master_clean.jpg_url.duplicated())


In [ ]:
fig, ax = plt.subplots(figsize = (20,7))
ax = sns.heatmap(df_master_clean.isnull(), vmin = 0, vmax = 1)

The last step here is to drop the rows which contain images, that are not displaying any dogs (relying on the top three predictions of the ML algorithm).


In [ ]:
df_master_clean.drop(df_master_clean.query("p1_dog == False and p2_dog == False and p3_dog == False").index, inplace = True)


In [ ]:
df_master_clean.query("p1_dog == False and p2_dog == False and p3_dog == False")

In [ ]:
df_master_clean.info()

<h4>3. Clean the datatypes of the columns

In [ ]:
#change the type of tweet_id to str:
df_master_clean["tweet_id"] = df_master_clean["tweet_id"].astype("str")

In [ ]:
#change   type of  timestamp to datetime
df_master_clean["timestamp"] = pd.to_datetime(df_master_clean.timestamp)


In [ ]:
for x in ["retweet_count", "favorite_count", "img_num"]:
    df_master_clean[x] = df_master_clean[x].astype("int64")

In [ ]:
df_master_clean.info()

<h4> 4. Clean the wrong numerators - the floats on the one hand, the ones with multiple occurence of the pattern on the other



While assessing the dataset, we found out, that floating numbers got transformed into integers, which lead to loss of information.


In [ ]:
tw_arch_copy.query("check_num == False")[["rating_numerator", "num","check_num"]]


We dropped a lot of rows, so we cannot be sure that all of these problems are still in this dataset, so we will extract it again.



In [ ]:
pattern = "(\d+\.\d+\/\d+)"

df_master_clean.text.str.extract(pattern, expand = True)[0].dropna()

In [ ]:
#get the right numerator out of the string
df_num_clean = df_master_clean.text.str.extract(pattern, expand = True)[0].dropna().str.split('/', n=1, expand=True)[0]

In [ ]:
df_num_clean

In [ ]:
#get the index of the wrong data
df_num_clean_index = df_num_clean.index
df_num_clean_values = df_num_clean.values.astype("float64")

Now that we have our data together, we can impute these values and clean this data.



In [ ]:
#transform the datatypes to float
df_master_clean.rating_numerator = df_master_clean.rating_numerator.astype("float64")
df_master_clean.rating_denominator = df_master_clean.rating_denominator.astype("float64")
#impute the data
df_master_clean.loc[df_num_clean_index, "rating_numerator"] = df_num_clean_values
df_master_clean.loc[df_num_clean_index].rating_numerator

We also have the problem, that there can be multiple occurrences of the pattern. The reason for this is - most of the time - the display of two or more dogs in an image. For this cases we could add the ratings up, because the author of the Twitter account did this in one case that we found. Or we could build the average rating per each picture. For now, we are going to drop them out of the dataframe.

In [ ]:
pattern = "(\d+(\.\d+)?\/\d+(\.\d+)?)"

print(df_master_clean.text.str.count(pattern)[df_master_clean.text.str.count(pattern) != 1])

#get the index of the rows which contains the pattern more than once
pattern_clean_index = df_master_clean.text.str.count(pattern)[df_master_clean.text.str.count(pattern) != 1].index

In [ ]:
df_master_clean.drop(pattern_clean_index, inplace = True)


In [ ]:
df_master_clean.info()

In [ ]:
#no more occurrences of the mentioned problem are left
print(df_master_clean.text.str.count(pattern)[df_master_clean.text.str.count(pattern) != 1])


<h4>5. Extract the source from html code



Right now the source column is not giving us any useful information while looking at it. Because the relevant information is always between two "> <", the information will be easy to extract.



In [ ]:
df_master_clean.head(2)


In [ ]:
#https://stackoverflow.com/questions/3075130/what-is-the-difference-between-and-regular-expressions
df_master_clean.source = df_master_clean.source.str.extract("\>(.*?)\<", expand = True)


In [ ]:
df_master_clean.iloc[:,:3].head(2)


In [ ]:
df_master_clean.source.value_counts()


<h4>6. Split the text range into two separate columns



In [ ]:
df_master_clean[["display_text_range"]].info()


In [ ]:
df_master_clean.display_text_range[1]


Since the display_text_range column is interpreted as list, we can simply split it by using list indexing.


In [ ]:
#get the lower text range at list index 0
df_master_clean["lower_text_range"] = df_master_clean["display_text_range"].apply(lambda x: x[0])


In [ ]:
#get the lower text range at list index 1
df_master_clean["upper_text_range"] = df_master_clean["display_text_range"].apply(lambda x: x[1])
df_master_clean.drop("display_text_range", axis = 1, inplace = True)

In [ ]:
df_master_clean[["lower_text_range", "upper_text_range"]].head()


<h4>7. Remove the "None" out of the doggo, floofer, pupper and puppo column and merge them into one column



We want to reduce the columns into one for an easier analysis. For that we have to remove the None with "" at first to concat the columns together and afterswards with np.nan, so we could easily exclude these rows from a specific analysis.



In [ ]:
#replace "None" with "" in each column
for x in ["doggo", "floofer", "pupper", "puppo"]:
    df_master_clean[x].replace("None", "", inplace = True)

In [ ]:
#https://stackoverflow.com/questions/19377969/combine-two-columns-of-text-in-dataframe-in-pandas-python
#concat the columns together
df_master_clean['dog_class'] = df_master_clean['doggo'].map(str) + df_master_clean[
    'floofer'].map(str) + df_master_clean['pupper'].map(str) + df_master_clean['puppo'].map(str)

In [449]:
df_master_clean.dog_class.value_counts()


pupper         165
doggo          54 
puppo          22 
floofer        8  
doggopupper    1  
Name: dog_class, dtype: int64

In [450]:
#replace the leftover "" with np.nan
df_master_clean["dog_class"].replace("", np.nan, inplace = True)

In [451]:
df_master_clean.dog_class.value_counts()


pupper         165
doggo          54 
puppo          22 
floofer        8  
doggopupper    1  
Name: dog_class, dtype: int64

In [452]:
#count the occurrences of the pattern and show the rows with count > 1
df_master_clean.text.str.count(r'(\bpuppo\b|\bdoggo\b|\bfloofer\b|\bpupper\b)')[
    df_master_clean.text.str.count(r'(\bpuppo\b|\bdoggo\b|\bfloofer\b|\bpupper\b)') > 1]


132     2
357     3
379     2
578     2
624     2
699     2
739     2
887     2
916     2
932     2
1136    2
1236    2
1266    2
1320    3
Name: text, dtype: int64

we can see there are cases, in which there were multiple classifications. Let's extract the classes from the text and see where the differences occur.

In [453]:
df_master_clean["dog_class_re"] = df_master_clean.text.str.extract(
    r'(\bpuppo\b|\bdoggo\b|\bfloofer\b|\bpupper\b)', expand = True)


In [454]:
#http://queirozf.com/entries/visualization-options-for-jupyter-notebooks
#show the full text
pd.set_option('display.max_colwidth', -1)

#find the differences of the extract
df_master_clean[["text","dog_class", "dog_class_re"]].dropna(subset = ["dog_class_re"]).query("dog_class != dog_class_re")

text  \
138  At first I thought this was a shy doggo, but it's actually a Rare Canadian Floofer Owl. Amateurs would confuse the two. 11/10 only send dogs https://t.co/TXdT3tmuYk   
357  Here we have Burke (pupper) and Dexter (doggo). Pupper wants to be exactly like doggo. Both 12/10 would pet at same time https://t.co/ANBpEYHaho                       
379  This is Bones. He's being haunted by another doggo of roughly the same size. 12/10 deep breaths pupper everything's fine https://t.co/55Dqe0SJNj                       
578  Meet Maggie &amp; Lila. Maggie is the doggo, Lila is the pupper. They are sisters. Both 12/10 would pet at the same time https://t.co/MYwR4DQKll                       
624  Please stop sending it pictures that don't even have a doggo or pupper in them. Churlish af. 5/10 neat couch tho https://t.co/u2c9c7qSg8                               
699  This is just downright precious af. 12/10 for both pupper and doggo https://t.co/o5J479bZUC                                                                            
739  Like father (doggo), like son (pupper). Both 12/10 https://t.co/pG2inLaOda                                                                                             

    dog_class dog_class_re  
138  floofer   doggo        
357  NaN       pupper       
379  pupper    doggo        
578  NaN       doggo        
624  NaN       doggo        
699  NaN       pupper       
739  NaN       doggo

The difference occurs in 8 cases. We can read through the text and extract the correct dog_class. Afterwards we can impute the correct classes into the column - for cases, in which there are multiple dogs in it, we will impute np.nan for consistency.

In [455]:
#191 puppo
#200 floofer
#531 two dogs
#575 pupper
#889 two dogs
#956 not classified by author
#1063 two dogs
#1113 two dogs

df_master_clean.loc[191, "dog_class"] = "puppo"
df_master_clean.loc[200, "dog_class"] = "floofer"
df_master_clean.loc[531, "dog_class"] = np.nan
df_master_clean.loc[575, "dog_class"] = "pupper"
df_master_clean.loc[889, "dog_class"] = np.nan
df_master_clean.loc[956, "dog_class"] = np.nan
df_master_clean.loc[1063, "dog_class"] = np.nan
df_master_clean.loc[1113, "dog_class"] = np.nan

In [456]:
#find the differences of the extract - worked
df_master_clean[["text","dog_class", "dog_class_re"]].dropna(subset = ["dog_class_re"]).query("dog_class != dog_class_re")


text  \
138  At first I thought this was a shy doggo, but it's actually a Rare Canadian Floofer Owl. Amateurs would confuse the two. 11/10 only send dogs https://t.co/TXdT3tmuYk   
357  Here we have Burke (pupper) and Dexter (doggo). Pupper wants to be exactly like doggo. Both 12/10 would pet at same time https://t.co/ANBpEYHaho                       
379  This is Bones. He's being haunted by another doggo of roughly the same size. 12/10 deep breaths pupper everything's fine https://t.co/55Dqe0SJNj                       
578  Meet Maggie &amp; Lila. Maggie is the doggo, Lila is the pupper. They are sisters. Both 12/10 would pet at the same time https://t.co/MYwR4DQKll                       
624  Please stop sending it pictures that don't even have a doggo or pupper in them. Churlish af. 5/10 neat couch tho https://t.co/u2c9c7qSg8                               
699  This is just downright precious af. 12/10 for both pupper and doggo https://t.co/o5J479bZUC                                                                            
739  Like father (doggo), like son (pupper). Both 12/10 https://t.co/pG2inLaOda                                                                                             
956  This is Cooper. He began to tear up when his bone was taken from him. 11/10 stay strong pupper https://t.co/qI8yvqKG02                                                 

    dog_class dog_class_re  
138  floofer   doggo        
357  NaN       pupper       
379  pupper    doggo        
578  NaN       doggo        
624  NaN       doggo        
699  NaN       pupper       
739  NaN       doggo        
956  NaN       pupper

In [457]:
#drop the columns out
df_master_clean.drop(["doggo", "floofer", "pupper", "puppo", "dog_class_re"], inplace = True, axis = 1)


KeyError: "['doggo' 'floofer' 'pupper' 'puppo'] not found in axis"

Similar to the multiple occurence of a pattern for the numerator, we should do the same check here.


In [ ]:
#count the occurrences of the pattern and show the rows with count > 1
df_master_clean[["text", "dog_class"]].loc[
    df_master_clean.text.str.count(r'(\bpuppo\b|\bdoggo\b|\bfloofer\b|\bpupper\b)')[
        df_master_clean.text.str.count(r'(\bpuppo\b|\bdoggo\b|\bfloofer\b|\bpupper\b)') > 1].index]


ok, now all of these look are correct.

<h4> 8. Remove the wrong names of name column

Here we will also replace the wrong names with np.nan.


In [ ]:
for x in ["None", "a", "by", "the"]:
    df_master_clean["name"].replace(x, np.nan, inplace = True)

In [ ]:
df_master_clean.name.value_counts()


In [ ]:
df_master_clean.head(5)


In [ ]:
fig, ax = plt.subplots(figsize = (20,7))
ax = sns.heatmap(df_master_clean.isnull(), vmin = 0, vmax = 1)

In [ ]:
df_master_clean.info()

<h4>9. Reduce the prediction columns into two - breed and conf


In the next step we want to reduce the prediction columns into two - breed and confidence. The columns are already sorted by confidence. We will take the most likely prediction for each row which is supposed to be a dog.


In [ ]:
df_master_clean.query("p2_conf > p1_conf")

In [ ]:
df_master_clean.query("p3_conf > p1_conf")

In [ ]:
df_master_clean.query("p3_conf > p2_conf")

order is correct.



In [ ]:
#extract the most likely prediction which is a dog
def get_attr(x):
    """
    INPUT: 
        Columns in this order: Check1, Result1, Check2, Result2, Result3
    OUTPUT:
        Results based on the check in this columns
    """
    if x[0] == True:
        return x[1]
    elif x[2] == True:
        return x[3]
    else:
        return x[4]
    
df_master_clean["breed"] = df_master_clean[["p1_dog", "p1", "p2_dog", "p2", "p3"]].apply(get_attr, axis = 1)
df_master_clean["conf"] = df_master_clean[["p1_dog", "p1_conf", "p2_dog", "p2_conf", "p3_conf"]].apply(get_attr, axis = 1)


In [ ]:
df_master_clean.iloc[:, 12:]


In [ ]:
#drop the reduced columns
df_master_clean.drop(df_master_clean.columns[12:21], inplace = True, axis = 1)


In [ ]:
df_master_clean.head()


<h4> 10. Clean the new breed column by replacing the "_" with a whitespace and make them all lowercase

Now that we have our reduced column, we have to clean it for consistency.


In [ ]:
#replace "_" with " "
df_master_clean.breed = df_master_clean.breed.str.replace("_", " ")


In [ ]:
df_master_clean.breed

In [ ]:
#https://stackoverflow.com/questions/22245171/how-to-lowercase-a-python-dataframe-string-column-if-it-has-missing-values
#lower the strings
df_master_clean.breed = df_master_clean.breed.str.lower()

In [ ]:
df_master_clean.breed.value_counts().head(10)

In [ ]:
#reset index to match with the real amount of rows
df_master_clean.reset_index(drop = True, inplace = True)


In [ ]:
fig, ax = plt.subplots(figsize = (20,7))
ax = sns.heatmap(df_master_clean.isnull(), vmin = 0, vmax = 1)


In [ ]:
df_master_clean.info()


In [ ]:
df_master_clean[["breed", "conf"]].head(5)


In [ ]:
#save the data to a *.csv file
df_master_clean.to_csv('twitter_archive_master.csv', index = False)

<h3>Data Analysis</h3> <br>

Questions:<br>
    <ol>
<li> Based on the predicted, most likely dog breed: Which breed gets retweeted and favorited the most overall?
<li>How did the account develop (speaking about number of tweets, retweets, favorites, image number and length of the tweets)?
<li>Is there a pattern visible in the timing of the tweets?
</ol>



<h4> 1. Based on the predicted, most likely dog breed: Which breed gets retweeted and favorited the most overall?</h4>

 for answer this question we will first take a look on the frequency of the breed occurence and afterwards we will create a groupby object to sum up the favorite and retweet count of each breed in this dataset.



In [ ]:
#read in the master csv
df = pd.read_csv("twitter_archive_master.csv")

In [ ]:
df.head()

In [ ]:
df.columns

In [ ]:
#https://stackoverflow.com/questions/32891211/limit-the-number-of-groups-shown-in-seaborn-countplot
fig, ax = plt.subplots(figsize = (16,5))
ax = sns.countplot(x = "breed", data = df, order=df.breed.value_counts().iloc[:10].index, palette = "viridis")
ax.set_title("count of classified breeds in the dataset");
ax.set_ylim(0, 170)
#https://stackoverflow.com/questions/33179122/seaborn-countplot-with-frequencies
for p in ax.patches:
    ax.annotate('{:.0f}'.format(p.get_height()), (p.get_x()+0.3, p.get_height()+2))

The dogs displayed in the images are mostly golden retrievers with a count of 154 or labrador retrievers with a count of 105.



In [ ]:
df_breed_group = df[["retweet_count", "favorite_count", "breed"]].groupby("breed", as_index = False).sum()
df_breed_group.sample(5)

In [ ]:
df_breed_group.sort_values("retweet_count", ascending = False).head(10)


In [ ]:
df_breed_group.sort_values("favorite_count", ascending = False).head(10)

The golden retriever and the labrador retriever therefore also lead the list of most favorite and retweets.

In [ ]:
df_breed_group["sum"] = df_breed_group["retweet_count"] + df_breed_group["favorite_count"]

In [ ]:
df_breed_group.sort_values("sum", ascending = False).head(10)

In [ ]:
fig, ax = plt.subplots(figsize = (16,5))
ax = sns.barplot(x = "breed", y = "sum", data = df_breed_group.sort_values("sum", ascending=False).iloc[:10], palette = "viridis")
ax.set_title("sum of favorites and retweets per breed");


Now let's look at the most retweetet and favorited single tweet.



In [ ]:
df[["retweet_count", "favorite_count", "breed"]].sort_values("retweet_count", ascending = False).head(5)


In [ ]:
df[["retweet_count", "favorite_count", "breed"]].sort_values("favorite_count", ascending = False).head(5)


We can see that the most liked and retweeted tweet is in fact a labrador retriever, with golden retrievers not even being in the list. Let's see if there are big differences in the average rating.

In [ ]:
df_breed_group_mean = df[["rating_numerator", "breed"]].groupby("breed", as_index = False).mean()

In [ ]:
fig, ax = plt.subplots(figsize = (16,5))
ax = sns.barplot(x = "breed", y = "rating_numerator", data = df_breed_group_mean.sort_values("rating_numerator", ascending = False).iloc[:100], palette = "viridis")
ax.set_title("average rating_numerator per breed");
#https://code.i-harness.com/de/q/2135a8
ax.xaxis.set_ticklabels([])
plt.tight_layout()

While the most breeds are on average on nearly the same level of rating, there is some outlier visible.

In [ ]:
df_breed_group_mean.sort_values("rating_numerator", ascending = False).head(10)


In [ ]:
df[["breed", "rating_numerator"]].sort_values("rating_numerator", ascending = False).head(5)


In [ ]:
len(df.query("breed == 'labrador retriever'"))

In [ ]:
len(df.query("breed == 'soft-coated wheaten terrier'"))

In [ ]:
df.query("breed != 'soft-coated wheaten terrier'").rating_numerator.mean()


The soft-coated wheaten terrier got a very high mean rating. In fact, the labrador retriever got overall the biggest rating with 165, but since there are a lot more tweets with labrador retriever than for the soft-coated wheaten terrier, the one big rating of the soft-coated wheaten terrier has a higher weight then the one of the labrador retriever (14 tweets of soft-coated wheaten terrier and 105 of labrador retriever).

Not taking the outlier into account, this leads us to an average rating of 11. Based on the number of posts, retweets, favorites and mean rating, we will give the title of "Most overall liked dog of this Twitter account and its community" to the labrador retriever.



<h4>2. How did the account develop (speaking about number of tweets, retweets, favorites, image number and length of the tweets)?</h4>

for answer this question we first have to extract time information out of the timestamp.



In [ ]:
df_time = df.copy()

In [ ]:
df_time.info()


The timestamp is a string again, so I have to transform it again.

In [ ]:
df_time.timestamp = pd.to_datetime(df_time.timestamp)

Let's first take a look on the day of the week.

In [ ]:
df_time["dow"] = df_time["timestamp"].apply(lambda x: x.dayofweek)


In [ ]:
fig, ax = plt.subplots(figsize = (16,5))
ax = sns.countplot(x = "dow", data = df_time, palette = "viridis")
ax.set_title("count of tweets in the dataset for each day of the week");

ax.set_ylim(0, 300)
#https://stackoverflow.com/questions/33179122/seaborn-countplot-with-frequencies
for p in ax.patches:
    ax.annotate('{:.0f}'.format(p.get_height()), (p.get_x()+0.3, p.get_height()+5))

We can see that most of the tweets are posted on monday. For tuesday to friday it is nearly the same number of posts. On the weekend the Twitter profile tweets a little bit less.



In [ ]:
df_time.timestamp.min()


In [ ]:
df_time.timestamp.max()


This dataset contains data from the end of 2015 to the August of 2017. Let's extract the month, year and hour information from the timestamp.



In [ ]:
#get the month out of the timestamp
df_time["month"] = df_time["timestamp"].apply(lambda x: x.month)
#get the year out of the timestamp
df_time["year"] = df_time["timestamp"].apply(lambda x: x.year)
#get the hour out of the timestamp
df_time["hour"] = df_time["timestamp"].apply(lambda x: x.hour)

For the first graph I only want to take a look on the full year 2016.


In [ ]:
fig, ax = plt.subplots(figsize = (16,5))
ax = sns.countplot(x = "month", data = df_time.query("year == 2016"), palette = "viridis")
ax.set_title("count of tweets in the dataset for each month 2016");

ax.set_ylim(0,150)
#https://stackoverflow.com/questions/33179122/seaborn-countplot-with-frequencies
for p in ax.patches:
    ax.annotate('{:.0f}'.format(p.get_height()), (p.get_x()+0.3, p.get_height()+1.5))



Over the timeperiod of 2016 the number of post per months decreased. It went from 134 tweets in January to 52 in December. Does this mean, that the performance of this account is also decreasing?

In [ ]:
#create a timestamp containing month and year
#https://erikrood.com/Python_References/extract_month_year_pandas_final.html
df_time['month_year'] = pd.to_datetime(df["timestamp"]).dt.to_period('M')


In [ ]:
ig, ax = plt.subplots(figsize = (16,5))
ax = sns.countplot(x = "month_year", data = df_time.sort_values("month_year"), palette = "viridis", )
ax.set_title("count of tweets in the dataset for each year - month combination in this dataset");

ax.set_ylim(0, 300)
#https://stackoverflow.com/questions/33179122/seaborn-countplot-with-frequencies
for p in ax.patches:
    ax.annotate('{:.0f}'.format(p.get_height()), (p.get_x()+0.2, p.get_height()+3))
plt.tight_layout()


If we look at it over the whole timeperiod it becomes even more clear. In April 2016 the number of tweets dropped and since then it has a relatively stable level. To see if the performance of the Account decreased we will take a look on the favorites and retweets that the posts get.



In [ ]:
df_time_groupby = df_time.groupby("month_year", as_index = False).sum()

df_time_groupby.head()


In [ ]:
fig, ax = plt.subplots(figsize = (16,5))
ax = sns.barplot(x = "month_year", y = "favorite_count", data = df_time_groupby, palette = "viridis")
ax.set_title("sum of favorites per month-year combination");
plt.tight_layout()


In [ ]:
fig, ax = plt.subplots(figsize = (16,5))
ax = sns.barplot(x = "month_year", y = "retweet_count", data = df_time_groupby, palette = "viridis")
ax.set_title("sum of retweets per month-year combination");


Interesting, while the number of tweets per month is decreasing, the favorites and retweets per month are increasing.


In [ ]:
#create a groupby object over the mean
df_time_groupby_mean = df_time.groupby("month_year", as_index = False).mean()

In [ ]:
fig, ax = plt.subplots(figsize = (16,5))
ax = sns.barplot(x = "month_year", y = "favorite_count", data = df_time_groupby_mean, palette = "viridis")
ax.set_title("mean of favorites per month-year combination");
plt.tight_layout()

In [ ]:
fig, ax = plt.subplots(figsize = (16,5))
ax = sns.barplot(x = "month_year", y = "retweet_count", data = df_time_groupby_mean, palette = "viridis")
ax.set_title("mean of retweets per month-year combination");
plt.tight_layout()

If we look at the average number of favorites and retweets the clear uptrend gets even more clearclearer! Now let's see if the number of posted images per month or the average upper text range changed over time.

In [ ]:
fig, ax = plt.subplots(figsize = (16,5))
ax = sns.barplot(x = "month_year", y = "img_num", data = df_time_groupby_mean, palette = "viridis")
ax.set_title("mean image number per month-year combination");
plt.tight_layout()

For the images it seems pretty stable. There are months where are more and months where are less posted images, but overall there is no clean up- or downtrend visible.



In [ ]:
fig, ax = plt.subplots(figsize = (16,5))
ax = sns.barplot(x = "month_year", y = "upper_text_range", data = df_time_groupby_mean, palette = "viridis")
ax.set_title("mean upper text range per month_year combination");

In [ ]:
df_time_groupby_mean.upper_text_range.mean()


In [ ]:
df_time_groupby_mean.iloc[:11].upper_text_range.mean()


In [ ]:
df_time_groupby_mean.iloc[11:].upper_text_range.mean()


For the tweet length it seems like it increased over the second half of this dataset from an average of 106 to 113.



<h4>3. Is there a pattern visible in the timing of the tweets?



In [ ]:
fig, ax = plt.subplots(figsize = (16,5))
ax = sns.countplot(x = "hour", data = df_time.query("year == 2016"), palette = "viridis")
ax.set_title("count of tweets in the dataset for each hour in 2016");


In [ ]:
fig, ax = plt.subplots(figsize = (16,5))
ax = sns.countplot(x = "hour", data = df_time.query("year == 2015"), palette = "viridis")
ax.set_title("count of tweets in the dataset for each hour in 2015");


we can see in this graphs of 2015 and 2016, the most posts are during the night between 0 (inlcuded) and 5 'o clock. between 4 and 15 'o clock there is a very small amounts of tweets, in fact there is no single tweet for the hours 7 - 12 and 14 o' clock. There are some tweets after 14 'o clock, but not as much as between 0 (included) and 5 'o clock.



<h2> Summary and Conclusions </h2>

In this project we went through a whole data wrangling process. The gathering included the programmatical, manual download as also the gathering via an API. Also the cleaning part covered a lot of techniques like regular expressions or the overall handling of dataframes.



In the analysis we took a look at following questions:



<h3> Questions: </h3>
<ol>
<li>Based on the predicted, most likely dog breed: Which breed gets retweeted and favorited the most overall?
The winner for our analysis was the labrador retriever.
<li>How did the account develop (speaking about number of tweets, retweets, favorites, image number and length of the tweets)?
We found, that the number of tweets per month decreased, while the retweets and favorites show an uptrend. For the image numbers there is no clear trend visible, the length of the tweets got a little bit closer to the maximum of 130 in the second half of the dataset.
<li>Is there a pattern visible in the timing of the tweets?
Here we found, that between 5 and 15 'o clock there are nearly no tweets at all. The most tweets are during the time from 0 - 4 and then from 15 - 23, but from 15 - 23 are less tweets than between 0 - 4.


